### PR TITLE
GH-13533 : Change channel rename command to be uniform with team rename

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
This PR updates the existing `Channel Rename` command to follow similar structure as `Team Rename`

#### Ticket Link
Fixes : https://github.com/mattermost/mattermost-server/issues/13533
Jira : None

#### Tests
Tests are intentionally not included so as to be taken at https://github.com/mattermost/mattermost-server/issues/13462 issue

### CODE REVIEW

**Comment by agnivade** [comment]
Apologies for the delay on this @M-ZubairAhmed. I will get to this tomorrow.

File: commands/channel.go [hunk 1]
```diff
@@ -23,12 +23,16 @@ var ChannelCreateCmd = &cobra.Command{
 	RunE: withClient(createChannelCmdF),
 }
 
+// ChannelRenameCmd is the command in mmctl to change name and/or display name of existing channel.
```
Comment (agnivade) [inline commands/channel.go]: I would suggest: "ChannelRenameCmd is used to change name and/or display name of an existing channel."

File: commands/channel.go [hunk 2]
```diff
@@ -345,32 +350,35 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
 }
 
 func renameChannelCmdF(c client.Client, cmd *cobra.Command, args []string) error {
-	var newDisplayName, newChannelName string
+	existingTeamChannel := args[0]
+	newChannelName, _ := cmd.Flags().GetString("name")
```
Comment (agnivade) [inline commands/channel.go]: I see that rename team also ignores the errors. Do you know why ?

File: commands/channel.go [hunk 3]
```diff
@@ -345,32 +350,35 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
+	newDisplayName, _ := cmd.Flags().GetString("display_name")
 
-	if len(args) < 2 {
-		return errors.New("Not enough arguments.")
+	// Both Display and name flags are not entered
+	if newDisplayName == "" && newChannelName == "" {
+		return errors.New("Require atleast one flag to rename team, either 'name' or 'display_name'")
```
Comment (agnivade) [inline commands/channel.go]: Copy paste error: rename _channel_ :)

File: commands/channel.go [hunk 4]
```diff
@@ -345,32 +350,35 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
 
 func renameChannelCmdF(c client.Client, cmd *cobra.Command, args []string) error {
-	var newDisplayName, newChannelName string
+	existingTeamChannel := args[0]
+	newChannelName, _ := cmd.Flags().GetString("name")
+	newDisplayName, _ := cmd.Flags().GetString("display_name")
 
-	if len(args) < 2 {
-		return errors.New("Not enough arguments.")
+	// Both Display and name flags are not entered
```
Comment (agnivade) [inline commands/channel.go]: I would suggest: "Atleast one of the flags must be present."

File: commands/channel.go [hunk 5]
```diff
@@ -345,32 +350,35 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
-		channelPatch.DisplayName = &newDisplayName
+		patchedChannel.DisplayName = &newDisplayName
 	}
 
-	if _, response := c.PatchChannel(channel.Id, &channelPatch); response.Error != nil {
-		return response.Error
+	// Using PatchChannel API to rename channel
+	_, response := c.PatchChannel(channel.Id, patchedChannel)
+	if response.Error != nil {
+		return errors.New("Cannot rename team '" + existingTeamChannel + "', error : " + response.Error.Error())
```
Comment (agnivade) [inline commands/channel.go]: Same: rename _channel_. Also `existingTeamChannel` contains both team and channel names, so cannot rename channel is not really accurate there. Let's pass the `channel.Name` here instead of existingTeamChannel.

File: docs/mmctl_user_email.md [hunk 6]
```diff
@@ -13,8 +13,7 @@ mmctl user email [user] [new email] [flags]
 ### Examples
 
 ```
-  user email test user@example.com
-  user activate username
```
Comment (agnivade) [inline docs/mmctl_user_email.md]: This seems to be unrelated.

File: docs/mmctl_channel_rename.md [hunk 7]
```diff
@@ -4,23 +4,26 @@ Rename a channel
...
+  channel rename myteam:oldchannel --name 'new-channel'
+  channel rename myteam:oldchannel --display_name 'New Display Name'
 ```
 
 ### Options
 
 ```
       --display_name string   Channel Display Name
   -h, --help                  help for rename
+      --name string           Channel Name
```
Comment (agnivade) [inline docs/mmctl_channel_rename.md]: Let's move this above help.

File: commands/channel.go [hunk 8]
```diff
@@ -345,32 +350,35 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
 	}
 
+	printer.Print("'" + existingTeamChannel + "' channel renamed")
```
Comment (agnivade) [inline commands/channel.go]: Same as above: let's use `channel.Name`

File: commands/channel.go [hunk 9]
```diff
@@ -345,32 +350,35 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
+	// Both Display and name flags are not entered
+	if newDisplayName == "" && newChannelName == "" {
+		return errors.New("Require atleast one flag to rename team, either 'name' or 'display_name'")
 	}
 
-	channel := getChannelFromChannelArg(c, args[0])
+	channel := getChannelFromChannelArg(c, existingTeamChannel)
 	if channel == nil {
-		return errors.New("Unable to find channel '" + args[0] + "'")
+		return errors.New("Unable to find channel '" + existingTeamChannel + "'")
```
Comment (agnivade) [inline commands/channel.go]: I would suggest a slight modification to be more accurate: "Unable to find channel from ...". Because `existingTeamChannel` is not just the channel name.

File: commands/channel.go [hunk 10]
```diff
@@ -345,32 +350,35 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
 	}
 
-	newChannelName = args[1]
-	newDisplayName, errdn := cmd.Flags().GetString("display_name")
-	if errdn != nil {
-		return errdn
+	patchedChannel := &model.ChannelPatch{}
```
Comment (agnivade) [inline commands/channel.go]: I thought `channelPatch` as the variable name was better. It's not really a "patched" channel because the patch is going to happen and this is the patch. Can you explain your reasoning behind selecting this name ?

File: commands/channel.go (hunk 2 above)
Comment (M-ZubairAhmed) [reply commands/channel.go]: I am not able to find where but i think @mgdelacroix suggested that if there was an error in getting flag, then the flag variable would be empty and error will be thrown in the next code step. Anyways what do you suggest

File: docs/mmctl_user_email.md (hunk 6 above)
Comment (M-ZubairAhmed) [reply docs/mmctl_user_email.md]: this is because of `make docs`

File: docs/mmctl_channel_rename.md (hunk 7 above)
Comment (M-ZubairAhmed) [reply docs/mmctl_channel_rename.md]: this is autogenerated, i didnt do it manually

File: docs/mmctl_user_email.md (hunk 6 above)
Comment (agnivade) [reply docs/mmctl_user_email.md]: Ah okay.

File: commands/channel.go (hunk 2 above)
Comment (agnivade) [reply commands/channel.go]: I think that might clash with explicitly passing an empty value. In this case, we need a value, so this is fine. But in some cases, we might have a use case for passing empty value. In that case, that check is necessary. And I do see the error being checked in some parts of the codebase, so for the sake of consistency I suggest to add the check.

Comment (M-ZubairAhmed) [reply commands/channel.go]: got it

Comment (M-ZubairAhmed) [reply commands/channel.go]: i will add now

Comment (M-ZubairAhmed) [reply commands/channel.go]: But i also see in many places it is being checked to see if flag is being empty. What error message should it be here in our case ?

File: commands/channel.go [hunk 11]
```diff
@@ -345,32 +350,43 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
 
 func renameChannelCmdF(c client.Client, cmd *cobra.Command, args []string) error {
-	var newDisplayName, newChannelName string
+	existingTeamChannel := args[0]
 
-	if len(args) < 2 {
-		return errors.New("Not enough arguments.")
+	newChannelName, err := cmd.Flags().GetString("name")
+	if err != nil {
+		return errors.New("Cannot get name flag, error : " + err.Error())
```
Comment (agnivade) [inline commands/channel.go]: Let's just directly return the error. It will be of the form: `fmt.Errorf("flag accessed but not defined: %s", name)`. So the name is already there.

File: commands/channel.go [hunk 12]
```diff
@@ -345,32 +350,43 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
 	}
 
-	channel := getChannelFromChannelArg(c, args[0])
-	if channel == nil {
-		return errors.New("Unable to find channel '" + args[0] + "'")
+	newDisplayName, err := cmd.Flags().GetString("display_name")
+	if err != nil {
+		return errors.New("Cannot get display_name flag, error : " + err.Error())
```
Comment (agnivade) [inline commands/channel.go]: Same as above.

File: docs/mmctl_channel_rename.md (hunk 7 above)
Comment (agnivade) [reply docs/mmctl_channel_rename.md]: Oh it's alphabetical. Got it.

File: commands/channel.go [hunk 13]
```diff
@@ -345,32 +350,43 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
+		return err
 	}
 
-	newChannelName = args[1]
-	newDisplayName, errdn := cmd.Flags().GetString("display_name")
-	if errdn != nil {
-		return errdn
+	// Atleast one of display name or name flag must be present
+	if newDisplayName == "" && newChannelName == "" {
+		return errors.New("Require atleast one flag to rename channel, either 'name' or 'display_name'")
```
Comment (mgdelacroix) [inline commands/channel.go]: Typo:

```suggestion
		return errors.New("Require at least one flag to rename channel, either 'name' or 'display_name'")
```

File: commands/channel.go [hunk 14]
```diff
@@ -345,32 +350,43 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
+	newDisplayName, err := cmd.Flags().GetString("display_name")
+	if err != nil {
+		return err
 	}
 
-	newChannelName = args[1]
-	newDisplayName, errdn := cmd.Flags().GetString("display_name")
-	if errdn != nil {
-		return errdn
+	// Atleast one of display name or name flag must be present
```
Comment (mgdelacroix) [inline commands/channel.go]: Typo:

```suggestion
	// At least one of display name or name flag must be present
```

File: commands/channel.go [hunk 15]
```diff
@@ -345,32 +350,43 @@ func makeChannelPrivateCmdF(c client.Client, cmd *cobra.Command, args []string)
...
 
-	if _, response := c.PatchChannel(channel.Id, &channelPatch); response.Error != nil {
-		return response.Error
+	// Using PatchChannel API to rename channel
+	_, response := c.PatchChannel(channel.Id, channelPatch)
+	if response.Error != nil {
+		return errors.New("Cannot rename channel '" + channel.Name + "', error : " + response.Error.Error())
 	}
 
+	printer.Print("'" + channel.Name + "' channel renamed")
```
Comment (mgdelacroix) [inline commands/channel.go]: As this is an update action, we should print the updated channel in case the user requests the JSON format. So if we modify the patch line to get the updated channel:

```go
updatedChannel, response := c.PatchChannel(channel.Id, channelPatch)
```

We can modify the print to be:

```suggestion
	printer.PrintT("'{{.Name}}' channel renamed", updatedChannel)
```
